File: server/app/matcha/routes/legal_defense.py
"""Legal Pilot routes (`/legal-pilot`, feature `legal_defense`).

Litigation-readiness evidence assembly for full-platform (Pro) companies. An
admin opens a legal matter, converses with a grounded AI that organizes the
company's own records (service: `services/legal_defense.py`), and exports an
attorney-facing packet (memo PDF + ZIP bundle). Business-facing, tenant-isolated.
A token-gated public route delivers a generated packet to outside counsel.

The AI is an organizer, not an advocate (see the service). Every matter mutation,
chat turn, packet generation, download, and share is written to
`legal_matter_audit_log` (legal-grade trail).
"""

import json
import logging
import secrets
import unicodedata
from datetime import date, datetime, timedelta, timezone
from typing import Literal, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Request, Response
from fastapi.responses import StreamingResponse
from pydantic import BaseModel, Field, field_validator

from ...database import get_connection
from ..dependencies import require_admin_or_client, get_client_company_id
from app.core.feature_flags import merge_company_features
from app.core.services.redis_cache import check_rate_limit, client_ip
from app.core.services.storage import get_storage
from ..services import legal_defense as ld
from ..services import legal_research

logger = logging.getLogger(__name__)

router = APIRouter()
public_router = APIRouter()

_MATTER_TYPES = ("subpoena", "class_action", "eeoc_charge", "single_plaintiff", "audit", "other")


# --------------------------------------------------------------------------- #
# Models
# --------------------------------------------------------------------------- #

class _JurisdictionStateMixin(BaseModel):
    # check_fields=False: the field lives on the concrete models below, not
    # on this mixin — without it Pydantic v2 rejects the class at definition.
    @field_validator("jurisdiction_state", check_fields=False)
    @classmethod
    def _upper_state(cls, v):
        return v.upper() if v else v


class MatterCreate(_JurisdictionStateMixin):
    title: str = Field(..., min_length=1, max_length=300)
    matter_type: Literal["subpoena", "class_action", "eeoc_charge", "single_plaintiff", "audit", "other"] = "other"
    allegation: Optional[str] = Field(None, max_length=20_000)
    defense_theory: Optional[str] = Field(None, max_length=20_000)
    evidence_start: Optional[date] = None
    evidence_end: Optional[date] = None
    counsel_directed: bool = False
    counsel_name: Optional[str] = Field(None, max_length=200)
    counsel_email: Optional[str] = Field(None, max_length=320)
    location_id: Optional[UUID] = None
    jurisdiction_state: Optional[str] = Field(None, min_length=2, max_length=2)


class MatterUpdate(_JurisdictionStateMixin):
    title: Optional[str] = Field(None, min_length=1, max_length=300)
    allegation: Optional[str] = Field(None, max_length=20_000)
    defense_theory: Optional[str] = Field(None, max_length=20_000)
    evidence_start: Optional[date] = None
    evidence_end: Optional[date] = None
    status: Optional[Literal["draft", "active", "closed"]] = None
    counsel_directed: Optional[bool] = None
    counsel_name: Optional[str] = Field(None, max_length=200)
    counsel_email: Optional[str] = Field(None, max_length=320)
    location_id: Optional[UUID] = None
    jurisdiction_state: Optional[str] = Field(None, min_length=2, max_length=2)


class ResearchOptions(BaseModel):
    # False skips the ~90s grounded-Gemini synthesis call — just the
    # CourtListener case search, for a fast case-law-only refresh.
    include_guidance: bool = True


class ChatIn(BaseModel):
    message: str = Field(..., min_length=1, max_length=5_000)


class PacketIn(BaseModel):
    kind: Literal["pdf", "zip", "both"] = "both"
    include_research: bool = False


class ShareIn(BaseModel):
    recipient_email: Optional[str] = Field(None, max_length=320)
    expires_days: int = Field(14, ge=1, le=365)


# --------------------------------------------------------------------------- #
# Helpers
# --------------------------------------------------------------------------- #

async def _audit(conn, matter_id, current_user, request: Request, action: str, details: dict | None = None):
    await conn.execute(
        """INSERT INTO legal_matter_audit_log (matter_id, user_id, action, details, ip_address)
           VALUES ($1, $2, $3, $4, $5)""",
        matter_id, getattr(current_user, "id", None), action,
        json.dumps(details or {}), client_ip(request),
    )


async def _load_matter(conn, matter_id: str, company_id) -> dict:
    row = await conn.fetchrow(
        "SELECT * FROM legal_matters WHERE id = $1 AND company_id = $2",
        matter_id, company_id,
    )
    if not row:
        raise HTTPException(status_code=404, detail="Matter not found")
    return dict(row)


async def _check_location(conn, location_id, company_id) -> None:
    row = await conn.fetchrow(
        "SELECT 1 FROM business_locations WHERE id = $1 AND company_id = $2",
        location_id, company_id,
    )
    if not row:
        raise HTTPException(status_code=400, detail="Location not found")


async def _features(conn, company_id) -> dict:
    row = await conn.fetchrow(
        "SELECT enabled_features, signup_source FROM companies WHERE id = $1", company_id
    )
    if not row:
        return {}
    return merge_company_features(row["enabled_features"], row["signup_source"])


async def _load_messages(conn, matter_id: str) -> list[dict]:
    rows = await conn.fetch(
        "SELECT role, content, metadata, created_at FROM legal_matter_messages "
        "WHERE matter_id = $1 ORDER BY created_at",
        matter_id,
    )
    return [dict(r) for r in rows]


async def _latest_memo(conn, matter_id: str) -> Optional[dict]:
    row = await conn.fetchrow(
        "SELECT content, metadata FROM legal_matter_messages "
        "WHERE matter_id = $1 AND role = 'assistant' ORDER BY created_at DESC LIMIT 1",
        matter_id,
    )
    if not row:
        return None
    meta = row["metadata"]
    if isinstance(meta, str):
        try:
            meta = json.loads(meta)
        except Exception:
            meta = {}
    meta = meta or {}
    return {
        "assistant_text": row["content"] or "",
        "evidence_map": meta.get("evidence_map") or [],
        "open_questions": meta.get("open_questions") or [],
    }


# --------------------------------------------------------------------------- #
# Matters CRUD
# --------------------------------------------------------------------------- #

@router.post("/matters")
async def create_matter(body: MatterCreate, request: Request, current_user=Depends(require_admin_or_client)):
    company_id = await get_client_company_id(current_user)
    async with get_connection() as conn:
        if body.location_id:
            await _check_location(conn, body.location_id, company_id)
        row = await conn.fetchrow(
            """
            INSERT INTO legal_matters
                (company_id, title, matter_type, allegation, defense_theory,
                 evidence_start, evidence_end, counsel_directed, counsel_name,
                 counsel_email, created_by, location_id, jurisdiction_state, status)
            VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,'active')
            RETURNING *
            """,
            company_id, body.title, body.matter_type, body.allegation, body.defense_theory,
            body.evidence_start, body.evidence_end, body.counsel_directed,
            body.counsel_name, body.counsel_email, getattr(current_user, "id", None),
            body.location_id, body.jurisdiction_state,
        )
        await _audit(conn, row["id"], current_user, request, "create", {"title": body.title})
    return dict(row)


@router.get("/matters")
async def list_matters(current_user=Depends(require_admin_or_client)):
    company_id = await get_client_company_id(current_user)
    async with get_connection() as conn:
        rows = await conn.fetch(
            """
            SELECT m.*,
                   (SELECT COUNT(*) FROM legal_matter_packets p WHERE p.matter_id = m.id) AS packet_count
            FROM legal_matters m
            WHERE m.company_id = $1
            ORDER BY m.updated_at DESC
            """,
            company_id,
        )
    return [dict(r) for r in rows]


@router.get("/matters/{matter_id}")
async def get_matter(matter_id: str, current_user=Depends(require_admin_or_client)):
    company_id = await get_client_company_id(current_user)
    async with get_connection() as conn:
        matter = await _load_matter(conn, matter_id, company_id)
        matter["messages"] = await _load_messages(conn, matter_id)
        packets = await conn.fetch(
            "SELECT id, kind, filename, citations, file_size, generated_at "
            "FROM legal_matter_packets WHERE matter_id = $1 ORDER BY generated_at DESC",
            matter_id,
        )
        share_rows = await conn.fetch(
            """SELECT packet_id, recipient_email, download_count, last_downloaded_at,
                      expires_at, revoked, created_at
                 FROM legal_matter_share_links
                WHERE matter_id = $1 ORDER BY created_at DESC""",
            matter_id,
        )
        # Most recent share link per packet — a packet can be re-shared, we
        # only need the latest to answer "has counsel opened this".
        shares_by_packet: dict = {}
        for s in share_rows:
            pid = str(s["packet_id"])
            shares_by_packet.setdefault(pid, dict(s))
        matter["packets"] = [
            {**dict(p), "share": shares_by_packet.get(str(p["id"]))} for p in packets
        ]
    return matter


@router.patch("/matters/{matter_id}")
async def update_matter(matter_id: str, body: MatterUpdate, request: Request, current_user=Depends(require_admin_or_client)):
    company_id = await get_client_company_id(current_user)
    fields = body.model_dump(exclude_unset=True)
    if not fields:
        raise HTTPException(status_code=400, detail="No fields to update")
    async with get_connection() as conn:
        await _load_matter(conn, matter_id, company_id)  # ownership
        if fields.get("location_id"):
            await _check_location(conn, fields["location_id"], company_id)
        sets, vals = [], []
        for i, (k, v) in enumerate(fields.items(), start=1):
            sets.append(f"{k} = ${i}")
            vals.append(v)
        vals.append(matter_id)
        closed = ", closed_at = NOW()" if fields.get("status") == "closed" else ""
        await conn.execute(
            f"UPDATE legal_matters SET {', '.join(sets)}, updated_at = NOW(){closed} WHERE id = ${len(vals)}",
            *vals,
        )
        await _audit(conn, matter_id, current_user, request, "update", {"fields": list(fields.keys())})
        row = await _load_matter(conn, matter_id, company_id)
    return row


# --------------------------------------------------------------------------- #
# Evidence preview + grounded chat
# --------------------------------------------------------------------------- #

@router.get("/matters/{matter_id}/evidence")
async def get_evidence(matter_id: str, current_user=Depends(require_admin_or_client)):
    company_id = await get_client_company_id(current_user)
    async with get_connection() as conn:
        matter = await _load_matter(conn, matter_id, company_id)
        features = await _features(conn, company_id)
        corpus = await ld.gather_evidence(
            conn, company_id, matter["evidence_start"], matter["evidence_end"], features, matter=matter
        )
    # Return source summaries + counts (the flat index is internal).
    return {
        "sources": corpus["sources"],
        "notes": corpus["notes"],
        "total": sum(len(s["records"]) for s in corpus["sources"].values()),
        "legal_context": corpus.get("legal_context"),
    }


@router.post("/matters/{matter_id}/chat")
async def chat(matter_id: str, body: ChatIn, request: Request, current_user=Depends(require_admin_or_client)):
    company_id = await get_client_company_id(current_user)
    # Pre-work in one connection; release it before the long Gemini call.
    async with get_connection() as conn:
        matter = await _load_matter(conn, matter_id, company_id)
        history = await _load_messages(conn, matter_id)
        features = await _features(conn, company_id)
        corpus = await ld.gather_evidence(
            conn, company_id, matter["evidence_start"], matter["evidence_end"], features, matter=matter
        )
        await conn.execute(
            "INSERT INTO legal_matter_messages (matter_id, role, content) VALUES ($1, 'user', $2)",
            matter_id, body.message,
        )
        await _audit(conn, matter_id, current_user, request, "message", {"role": "user"})

    async def event_stream():
        result_payload = None
        try:
            async for ev in ld.run_chat_turn(matter, history, corpus, body.message):
                if ev.get("type") == "result":
                    result_payload = ev.get("data")
                yield f"data: {json.dumps(ev)}\n\n"
        except Exception:
            logger.exception("legal_defense: chat stream error")
            yield f"data: {json.dumps({'type': 'error', 'message': 'Analysis failed.'})}\n\n"
        # Persist the assistant turn (+ validated evidence map) after streaming.
        if result_payload:
            try:
                async with get_connection() as c2:
                    await c2.execute(
                        "INSERT INTO legal_matter_messages (matter_id, role, content, metadata) "
                        "VALUES ($1, 'assistant', $2, $3)",
                        matter_id, result_payload.get("assistant_text", ""),
                        json.dumps({
                            "evidence_map": result_payload.get("evidence_map"),
                            "open_questions": result_payload.get("open_questions"),
                            "dropped_citations": result_payload.get("dropped_citations"),
                        }),
                    )
                    await c2.execute(
                        "UPDATE legal_matters SET updated_at = NOW() WHERE id = $1", matter_id
                    )
            except Exception:
                logger.exception("legal_defense: failed to persist assistant message")
        yield "data: [DONE]\n\n"

    return StreamingResponse(
        event_stream(), media_type="text/event-stream", headers={"X-Accel-Buffering": "no"}
    )


# --------------------------------------------------------------------------- #
# External legal research (CourtListener case search + grounded Gemini
# guidance) — synchronous POST (worst case ~110s: 20s CourtListener + 90s
# Gemini), acceptable behind an explicit button. The row-status design
# supports a later move to BackgroundTasks + GET-poll with no schema change.
# --------------------------------------------------------------------------- #

@router.post("/matters/{matter_id}/research")
async def run_matter_research(
    matter_id: str, request: Request, body: ResearchOptions = ResearchOptions(),
    current_user=Depends(require_admin_or_client),
):
    company_id = await get_client_company_id(current_user)
    # Load (and 404) BEFORE consuming the rate budget — a typo'd matter id
    # must not burn the company's 10/hour research allowance.
    async with get_connection() as conn:
        matter = await _load_matter(conn, matter_id, company_id)
    await check_rate_limit(str(company_id), "legal_research", 10, 3600)
    # run_research manages its own short-lived connections so no pooled
    # connection is held across the ~110s of external CourtListener + Gemini
    # calls (same discipline as chat()'s pre-work/release pattern).
    row = await legal_research.run_research(
        matter, getattr(current_user, "id", None), include_guidance=body.include_guidance,
    )
    async with get_connection() as conn:
        await _audit(conn, matter_id, current_user, request, "research",
                     {"cases": len(row.get("cases") or []), "status": row.get("status"),
                      "include_guidance": body.include_guidance})
    return row


@router.get("/matters/{matter_id}/research")
async def list_matter_research(matter_id: str, current_user=Depends(require_admin_or_client)):
    company_id = await get_client_company_id(current_user)
    async with get_connection() as conn:
        await _load_matter(conn, matter_id, company_id)  # ownership
        rows = await conn.fetch(
            "SELECT * FROM legal_matter_research WHERE matter_id = $1 ORDER BY created_at DESC",
            matter_id,
        )
    return [legal_research.parse_research_row(dict(r)) for r in rows]


# --------------------------------------------------------------------------- #
# Packet generation + download + counsel share
# --------------------------------------------------------------------------- #

def _safe_name(s: str) -> str:
    """ASCII-only filename slug. Non-ASCII survives into Content-Disposition
    otherwise, and Starlette encodes header values as latin-1 — a title with
    an em dash or accent crashes every download of that packet with a 500."""
    ascii_s = unicodedata.normalize("NFKD", s or "matter").encode("ascii", "ignore").decode("ascii")
    return (ascii_s or "matter").replace("/", "-").replace('"', "").replace(" ", "-")[:60] or "matter"


def _safe_filename(name: Optional[str]) -> str:
    """Defensive re-sanitize for filenames already stored in the DB (packets
    created before ``_safe_name`` was hardened to strip non-ASCII)."""
    base, _, ext = (name or "packet").rpartition(".")
    return f"{_safe_name(base or name)}.{ext or 'bin'}"


@router.post("/matters/{matter_id}/packet")
async def generate_packet(matter_id: str, body: PacketIn, request: Request, current_user=Depends(require_admin_or_client)):
    company_id = await get_client_company_id(current_user)
    async with get_connection() as conn:
        matter = await _load_matter(conn, matter_id, company_id)
        memo = await _latest_memo(conn, matter_id)
        if not memo:
            raise HTTPException(status_code=400, detail="Discuss the matter in chat first — the packet is built from that analysis.")
        features = await _features(conn, company_id)
        corpus = await ld.gather_evidence(
            conn, company_id, matter["evidence_start"], matter["evidence_end"], features, matter=matter
        )
        company = await conn.fetchrow("SELECT name FROM companies WHERE id = $1", company_id)

        research_row = None
        if body.include_research:
            # Same jurisdiction guard as _gather_case_law: a run grounded in a
            # different state than the matter's CURRENT jurisdiction is stale
            # and must not ride along in the packet.
            current_state = (corpus.get("legal_context") or {}).get("state")
            research_row = await conn.fetchrow(
                "SELECT * FROM legal_matter_research WHERE matter_id = $1 AND status = 'complete' "
                "AND ($2::varchar IS NULL OR jurisdiction_state IS NULL OR jurisdiction_state = $2) "
                "ORDER BY created_at DESC LIMIT 1",
                matter_id, current_state,
            )
            research_row = legal_research.parse_research_row(dict(research_row)) if research_row else None

        packet = await ld.build_defense_packet(conn, matter, corpus, memo,
                                                company_name=company["name"] if company else None,
                                                research=research_row)

        storage = get_storage()
        base = _safe_name(matter.get("title"))
        out = []

        async def _store(kind: str, blob: bytes, ext: str, mime: str):
            path = await storage.upload_private_file(
                blob, f"legal-defense-{base}.{ext}", prefix="legal-defense", content_type=mime
            )
            row = await conn.fetchrow(
                """INSERT INTO legal_matter_packets
                       (matter_id, company_id, kind, storage_path, filename, citations, file_size, generated_by)
                   VALUES ($1,$2,$3,$4,$5,$6,$7,$8) RETURNING id, kind, filename, file_size, generated_at""",
                matter_id, company_id, kind, path, f"legal-defense-{base}.{ext}",
                json.dumps(packet["citations"]), len(blob), getattr(current_user, "id", None),
            )
            out.append(dict(row))

        if body.kind in ("pdf", "both"):
            await _store("pdf", packet["pdf"], "pdf", "application/pdf")
        if body.kind in ("zip", "both") and packet.get("zip"):
            await _store("zip", packet["zip"], "zip", "application/zip")

        await _audit(conn, matter_id, current_user, request, "generate_packet",
                     {"kind": body.kind, "citations": len(packet["citations"]), "research": body.include_research})
    return {"packets": out}


async def _owned_packet(conn, matter_id: str, packet_id: str, company_id) -> dict:
    row = await conn.fetchrow(
        """SELECT p.* FROM legal_matter_packets p
           JOIN legal_matters m ON m.id = p.matter_id
           WHERE p.id = $1 AND p.matter_id = $2 AND m.company_id = $3""",
        packet_id, matter_id, company_id,
    )
    if not row:
        raise HTTPException(status_code=404, detail="Packet not found")
    return dict(row)


@router.get("/matters/{matter_id}/packets/{packet_id}/download")
async def download_packet(matter_id: str, packet_id: str, request: Request, current_user=Depends(require_admin_or_client)):
    company_id = await get_client_company_id(current_user)
    async with get_connection() as conn:
        pkt = await _owned_packet(conn, matter_id, packet_id, company_id)
        await _audit(conn, matter_id, current_user, request, "export", {"packet_id": packet_id})
    data = await get_storage().download_file(pkt["storage_path"])
    mime = "application/zip" if pkt["kind"] == "zip" else "application/pdf"
    return Response(
        content=data, media_type=mime,
        headers={"Content-Disposition": f'attachment; filename="{_safe_filename(pkt["filename"])}"'},
    )


@router.post("/matters/{matter_id}/packets/{packet_id}/share")
async def share_packet(matter_id: str, packet_id: str, body: ShareIn, request: Request, current_user=Depends(require_admin_or_client)):
    company_id = await get_client_company_id(current_user)
    token = secrets.token_urlsafe(32)
    expires = datetime.now(timezone.utc) + timedelta(days=body.expires_days)
    async with get_connection() as conn:
        await _owned_packet(conn, matter_id, packet_id, company_id)  # ownership
        await conn.execute(
            """INSERT INTO legal_matter_share_links
                   (matter_id, company_id, packet_id, token, recipient_email, expires_at, created_by)
               VALUES ($1,$2,$3,$4,$5,$6,$7)""",
            matter_id, company_id, packet_id, token, body.recipient_email, expires,
            getattr(current_user, "id", None),
        )
        await _audit(conn, matter_id, current_user, request, "share",
                     {"packet_id": packet_id, "recipient": body.recipient_email})
    return {"token": token, "path": f"/legal-pilot/share/{token}", "expires_at": expires}


# --------------------------------------------------------------------------- #
# Public counsel download (no auth — token-gated). Mounted without the feature
# gate, like inbound_email's /report token routes.
# --------------------------------------------------------------------------- #

@public_router.get("/legal-pilot/share/{token}")
async def download_shared_packet(token: str, request: Request):
    await check_rate_limit(client_ip(request), "legal_share_dl", 30, 3600)
    async with get_connection() as conn:
        row = await conn.fetchrow(
            """SELECT s.id, s.revoked, s.expires_at, p.storage_path, p.filename, p.kind, s.matter_id
               FROM legal_matter_share_links s
               JOIN legal_matter_packets p ON p.id = s.packet_id
               WHERE s.token = $1""",
            token,
        )
        if not row:
            raise HTTPException(status_code=404, detail="Invalid or expired link")
        if row["revoked"]:
            raise HTTPException(status_code=410, detail="This link has been revoked")
        if row["expires_at"] is not None and row["expires_at"] <= datetime.now(timezone.utc):
            raise HTTPException(status_code=410, detail="This link has expired")
        await conn.execute(
            "UPDATE legal_matter_share_links SET download_count = download_count + 1, "
            "last_downloaded_at = NOW() WHERE id = $1",
            row["id"],
        )
        await conn.execute(
            "INSERT INTO legal_matter_audit_log (matter_id, action, details, ip_address) VALUES ($1,$2,$3,$4)",
            row["matter_id"], "shared_download", json.dumps({"token_id": str(row["id"])}), client_ip(request),
        )
    data = await get_storage().download_file(row["storage_path"])
    mime = "application/zip" if row["kind"] == "zip" else "application/pdf"
    return Response(
        content=data, media_type=mime,
        headers={"Content-Disposition": f'attachment; filename="{_safe_filename(row["filename"])}"'},
    )

File: client/src/api/hr/legalDefense.ts
// Legal Pilot API client. Matters CRUD + grounded chat (SSE) + packet
// generation/download/share. The chat stream is a raw fetch (like
// IRCopilotPanel) since api/client.ts doesn't stream; everything else goes
// through the typed `api` helper.
import { api, authStreamHeaders } from '../client'

const BASE = import.meta.env.VITE_API_URL ?? '/api'

export type MatterType = 'subpoena' | 'class_action' | 'eeoc_charge' | 'single_plaintiff' | 'audit' | 'other'
export type MatterStatus = 'draft' | 'active' | 'closed'

export type EvidenceRecord = { cid: string; ref: string | null; summary: string; when: string; when_iso?: string | null }
export type EvidenceSource = { label: string; records: EvidenceRecord[] }
export type JurisdictionChainLink = { id: string; level: string; display_name: string }
export type LegalContext = {
  jurisdiction_id: string
  chain: JurisdictionChainLink[]
  state: string | null
  location_name: string | null
}
/** Values accepted by `legal_matters.subject_theory`. `null` derives the subject
 *  from the allegation; `'all'` forces every subject into the corpus. */
export type SubjectTheory = 'wage_hour' | 'eeo' | 'safety' | 'all'

/** The subject the backend scoped evidence to — derived from the matter's
 *  allegation text, or read from its stored `subject_theory` override.
 *  Null = no subject filter applied. */
export type MatterTheory = { slug: string; label: string; overridden: boolean }

export type EvidencePreview = {
  sources: Record<string, EvidenceSource>
  notes: string[]
  total: number
  legal_context?: LegalContext | null
  theory?: MatterTheory | null
}

/** The sources a subject filter actually narrows. Training, policy
 *  acknowledgments and accommodations are deliberately never subject-filtered
 *  (they're the exculpatory half of the record), so an empty one of those must
 *  not be blamed on the subject. Mirrors `_THEORIES` in the service. */
export const SUBJECT_FILTERED_SOURCES = new Set([
  'incidents', 'er_cases', 'compliance', 'compliance_alerts', 'discipline',
])

export type EvidenceMapItem = { point: string; cited_ids: string[] }
export type MessageMeta = {
  evidence_map?: EvidenceMapItem[]
  open_questions?: string[]
  dropped_citations?: string[]
} | null

export type MatterMessage = {
  role: 'user' | 'assistant' | 'system'
  content: string
  metadata: MessageMeta
  created_at: string
}

export type PacketShareStatus = {
  recipient_email: string | null
  download_count: number
  last_downloaded_at: string | null
  expires_at: string | null
  revoked: boolean
  created_at: string
}

export type Packet = {
  id: string
  kind: 'pdf' | 'zip'
  filename: string
  citations?: string[]
  file_size: number | null
  generated_at: string
  share?: PacketShareStatus | null
}

export type Matter = {
  id: string
  company_id: string
  title: string
  matter_type: MatterType
  allegation: string | null
  defense_theory: string | null
  status: MatterStatus
  evidence_start: string | null
  evidence_end: string | null
  counsel_directed: boolean
  counsel_name: string | null
  counsel_email: string | null
  location_id: string | null
  jurisdiction_state: string | null
  /** null = derive the evidence subject from the allegation. */
  subject_theory: SubjectTheory | null
  response_deadline: string | null
  deadline_note: string | null
  created_at: string
  updated_at: string
  closed_at: string | null
  packet_count?: number
  messages?: MatterMessage[]
  packets?: Packet[]
}

export type MatterCreate = {
  title: string
  matter_type: MatterType
  allegation?: string | null
  defense_theory?: string | null
  /** null (or omitted) derives the evidence subject from the allegation. */
  subject_theory?: SubjectTheory | null
  evidence_start?: string | null
  evidence_end?: string | null
  counsel_directed?: boolean
  counsel_name?: string | null
  counsel_email?: string | null
  location_id?: string | null
  jurisdiction_state?: string | null
  response_deadline?: string | null
  deadline_note?: string | null
}

export type IntakeDraft = {
  matter_type: MatterType
  title: string | null
  allegation: string | null
  plaintiff: string | null
  defendant: string | null
  jurisdiction_state: string | null
  evidence_start: string | null
  evidence_end: string | null
  response_deadline: string | null
  available: boolean
}
export const parseIntakeDocument = (file: File) => {
  const fd = new FormData()
  fd.append('file', file)
  return api.upload<IntakeDraft>('/legal-pilot/intake/parse', fd)
}

export const listMatters = () => api.get<Matter[]>('/legal-pilot/matters')
export const createMatter = (body: MatterCreate) => api.post<Matter>('/legal-pilot/matters', body)
export const getMatter = (id: string) => api.get<Matter>(`/legal-pilot/matters/${id}`)
export const updateMatter = (id: string, body: Partial<MatterCreate> & { status?: MatterStatus }) =>
  api.patch<Matter>(`/legal-pilot/matters/${id}`, body)
export const getEvidence = (id: string) => api.get<EvidencePreview>(`/legal-pilot/matters/${id}/evidence`)
export const generatePacket = (id: string, kind: 'pdf' | 'zip' | 'both', includeResearch = false) =>
  api.post<{ packets: Packet[] }>(`/legal-pilot/matters/${id}/packet`, { kind, include_research: includeResearch })
export const sharePacket = (matterId: string, packetId: string, body: { recipient_email?: string; expires_days?: number }) =>
  api.post<{ token: string; path: string; expires_at: string }>(
    `/legal-pilot/matters/${matterId}/packets/${packetId}/share`, body,
  )

// Authed blob download → browser save (keeps the server filename).
// api.download handles proactive refresh + 401 retry.
export function downloadPacket(matterId: string, packet: Packet): Promise<void> {
  return api.download(
    `/legal-pilot/matters/${matterId}/packets/${packet.id}/download`,
    packet.filename,
  )
}

export type ResearchCase = {
  id: string
  case_name: string
  citation: string | null
  court: string
  date_filed: string | null
  url: string
  snippet?: string | null
}
export type ResearchGuidance = {
  summary: string
  key_authorities: { name: string; url: string; publisher?: string; relevance?: string }[]
}
export type ResearchRow = {
  id: string
  status: 'running' | 'complete' | 'failed'
  query: string | null
  jurisdiction_state: string | null
  cases: ResearchCase[] | null
  guidance: ResearchGuidance | null
  error: string | null
  created_at: string
  completed_at: string | null
}
export const runResearch = (matterId: string, includeGuidance = true) =>
  api.post<ResearchRow>(`/legal-pilot/matters/${matterId}/research`, { include_guidance: includeGuidance })
export const listResearch = (matterId: string) =>
  api.get<ResearchRow[]>(`/legal-pilot/matters/${matterId}/research`)

export type ChatResult = {
  assistant_text: string
  evidence_map: EvidenceMapItem[]
  open_questions: string[]
  dropped_citations?: string[]
}
export type ChatHandlers = {
  onStatus?: (message: string) => void
  onResult?: (data: ChatResult) => void
  onError?: (message: string) => void
}

// Grounded chat turn over SSE — raw fetch + ReadableStream (api/client.ts can't
// stream). Mirrors the IRCopilotPanel consumption pattern.
export async function streamChat(matterId: string, message: string, h: ChatHandlers): Promise<void> {
  // Streams can't replay a 401 refresh-and-retry — refresh proactively.
  const res = await fetch(`${BASE}/legal-pilot/matters/${matterId}/chat`, {
    method: 'POST',
    headers: await authStreamHeaders({ 'Content-Type': 'application/json' }),
    body: JSON.stringify({ message }),
  })
  if (!res.ok || !res.body) {
    h.onError?.(`Chat failed (${res.status})`)
    return
  }
  const reader = res.body.getReader()
  const decoder = new TextDecoder()
  let buf = ''
  while (true) {
    const { value, done } = await reader.read()
    if (done) break
    buf += decoder.decode(value, { stream: true })
    const events = buf.split('\n\n')
    buf = events.pop() || ''
    for (const ev of events) {
      if (!ev.startsWith('data: ')) continue
      const payload = ev.slice(6)
      if (payload === '[DONE]') return
      try {
        const data = JSON.parse(payload)
        if (data.type === 'status') h.onStatus?.(data.message)
        else if (data.type === 'result') h.onResult?.(data.data)
        else if (data.type === 'error') h.onError?.(data.message)
      } catch {
        /* ignore partial/non-JSON frames */
      }
    }
  }
}
